### PR TITLE
fix(reify): do not clean up omitted nodes with no parent

### PR DESCRIPTION
**Pull request by wraithgar**
Co-authored by @nlf

When reify goes to add omits to the trash list, sometimes it adds things
it shouldn't, such as top level nodes that are actually links, but
they're only links elsewhere in the tree where they are properly placed
(i.e. with the root as the parent).

This PR prevents things like "npm install --production" deleting the
contents of a linked dev dependency

## References
Closes https://github.com/npm/cli/issues/2889

### CODE REVIEW

**Comment by wraithgar** [comment]
Still needs a test to prevent regression on the bug it's fixing.

**Review by darcyclarke — CHANGES_REQUESTED** [review]
Waiting on test coverage

**Comment by isaacs** [comment]
Here's a test:

```patch
diff --git a/test/arborist/reify.js b/test/arborist/reify.js
index e3f9f19a..f58aa90c 100644
--- a/test/arborist/reify.js
+++ b/test/arborist/reify.js
@@ -1691,3 +1691,52 @@ console.log('ok 1 - this is fine')
     stdio: 'pipe',
   }), 'test result')
 })
+
+t.test('do not delete linked targets when link omitted', async t => {
+  const path = t.testdir({
+    'package.json': JSON.stringify({
+      devDependencies: {
+        foo: 'file:foo',
+      },
+      dependencies: {
+        abbrev: '',
+      },
+    }),
+    node_modules: {
+      foo: t.fixture('symlink', '../foo'),
+      abbrev: {
+        'package.json': JSON.stringify({
+          name: 'abbrev',
+          version: '1.2.3',
+        }),
+      },
+    },
+    foo: {
+      'package.json': JSON.stringify({
+        name: 'foo',
+        version: '1.2.3',
+        dependencies: {
+          bar: '1.2.3',
+        },
+      }),
+      'index.js': 'console.log("hello from foo")',
+      node_modules: {
+        bar: {
+          'package.json': JSON.stringify({
+            name: 'bar',
+            version: '1.2.3',
+          }),
+        },
+      },
+    },
+  })
+  const barpj = resolve(path, 'foo/node_modules/bar/package.json')
+  const fooindex = resolve(path, 'foo/index.js')
+  t.equal(fs.existsSync(barpj), true, 'bar package.json present')
+  t.equal(fs.existsSync(fooindex), true, 'foo index.js present')
+  const tree = await reify(path, { omit: ['dev'] })
+  t.equal(fs.existsSync(barpj), true, 'bar package.json still present')
+  t.equal(fs.existsSync(fooindex), true, 'foo index.js still present')
+  t.notOk(tree.children.get('foo'), 'does not have foo child any more')
+  t.equal(tree.fsChildren.size, 1, 'still has foo fschild')
+})

```

It looks like it's still deleting the _children_ of those linked targets.

File: lib/arborist/reify.js [hunk 1]
```diff
@@ -406,11 +406,14 @@ module.exports = cls => class Reifier extends cls {
...
 
     process.emit('time', 'reify:trashOmits')
+    // node.parent is checked to make sure this is a node that's in the tree, and
+    // not the parent-less top level nodes
     const filter = node =>
-      node.peer && this[_omitPeer] ||
-      node.dev && this[_omitDev] ||
-      node.optional && this[_omitOptional] ||
-      node.devOptional && this[_omitOptional] && this[_omitDev]
+      node.parent &&
```
Comment (isaacs) [inline lib/arborist/reify.js:412]:
```suggestion
      node.isDescendantOf(this.idealTree) &&
```

**Review by isaacs — CHANGES_REQUESTED** [review]
Should check that it's a descendant of the idealTree, rather than just that it's a top node.

Test:

```patch
diff --git a/lib/arborist/reify.js b/lib/arborist/reify.js
index 67bd2fd3..e62017f9 100644
--- a/lib/arborist/reify.js
+++ b/lib/arborist/reify.js
@@ -406,10 +406,11 @@ module.exports = cls => class Reifier extends cls {
       return
 
     process.emit('time', 'reify:trashOmits')
-    // node.parent is checked to make sure this is a node that's in the tree, and
-    // not the parent-less top level nodes
+    // Only trash nodes that are in the idealTree's node_modules
+    // folder, ie are descendants of the idealTree.  Anything outside
+    // should be left alone.
     const filter = node =>
-      node.parent &&
+      node.isDescendantOf(this.idealTree) &&
         (node.peer && this[_omitPeer] ||
           node.dev && this[_omitDev] ||
           node.optional && this[_omitOptional] ||
diff --git a/test/arborist/reify.js b/test/arborist/reify.js
index e3f9f19a..f58aa90c 100644
--- a/test/arborist/reify.js
+++ b/test/arborist/reify.js
@@ -1691,3 +1691,52 @@ console.log('ok 1 - this is fine')
     stdio: 'pipe',
   }), 'test result')
 })
+
+t.test('do not delete linked targets when link omitted', async t => {
+  const path = t.testdir({
+    'package.json': JSON.stringify({
+      devDependencies: {
+        foo: 'file:foo',
+      },
+      dependencies: {
+        abbrev: '',
+      },
+    }),
+    node_modules: {
+      foo: t.fixture('symlink', '../foo'),
+      abbrev: {
+        'package.json': JSON.stringify({
+          name: 'abbrev',
+          version: '1.2.3',
+        }),
+      },
+    },
+    foo: {
+      'package.json': JSON.stringify({
+        name: 'foo',
+        version: '1.2.3',
+        dependencies: {
+          bar: '1.2.3',
+        },
+      }),
+      'index.js': 'console.log("hello from foo")',
+      node_modules: {
+        bar: {
+          'package.json': JSON.stringify({
+            name: 'bar',
+            version: '1.2.3',
+          }),
+        },
+      },
+    },
+  })
+  const barpj = resolve(path, 'foo/node_modules/bar/package.json')
+  const fooindex = resolve(path, 'foo/index.js')
+  t.equal(fs.existsSync(barpj), true, 'bar package.json present')
+  t.equal(fs.existsSync(fooindex), true, 'foo index.js present')
+  const tree = await reify(path, { omit: ['dev'] })
+  t.equal(fs.existsSync(barpj), true, 'bar package.json still present')
+  t.equal(fs.existsSync(fooindex), true, 'foo index.js still present')
+  t.notOk(tree.children.get('foo'), 'does not have foo child any more')
+  t.equal(tree.fsChildren.size, 1, 'still has foo fschild')
+})

```

**Review by nlf — APPROVED** [review]

**Review by nlf — APPROVED** [review]
changes applied and test fixed :+1:

**Pull request merged**